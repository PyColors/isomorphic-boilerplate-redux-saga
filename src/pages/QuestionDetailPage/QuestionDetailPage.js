/**
 * This page return the list of the question collection
 * and takes: title, body, answerCount, tags from mapStateToProps
 */

import React from 'react';
import Markdown from 'react-markdown';
import { connect } from 'react-redux';
import TagsList from '../../components/molecules/TagsList';

export const QuestionDetailDisplay = ({ title, body, answerCount, tags }) => (
  <div>
    <h3 className="mb-2">{title}</h3>
    {body ? (
      <div>
        <div className="mb-3">
          <TagsList tags={tags} />
        </div>
        <Markdown source={body} />
        <div>{answerCount} Answers</div>
      </div>
    ) : (
      <div>
        <h4>Loading Question...</h4>
      </div>
    )}
  </div>
);

export const mapStateToProps = (state, ownProps) => ({
  ...state.question.find(
    /* eslint-disable-next-line */
    ({ question_id }) => question_id == ownProps.question_id
  )
});

export default connect(mapStateToProps)(QuestionDetailDisplay);
