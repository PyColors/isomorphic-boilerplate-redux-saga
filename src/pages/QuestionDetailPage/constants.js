/*
 * Questions list Constants
 *
 * With the following format:
 * export const ACTION_CONSTANT = 'appName/atomicComponentName/ACTION_CONSTANT';
 * Reducers involuntarily avoid picking up actions that they should not
 */

export const FETCHED_QUESTION =
  'boilerplate/QuestionDetailPage/FETCHED_QUESTIONS';

export const REQUEST_FETCH_QUESTION =
  'boilerplate/QuestionDetailPage/REQUEST_FETCH_QUESTION';
