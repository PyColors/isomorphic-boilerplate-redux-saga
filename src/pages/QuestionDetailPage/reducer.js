import unionWith from 'lodash/unionWith';

/**
 * Manage how to render the state of questions
 * @param state
 * @param type
 * @param question
 * @returns Array state
 */
export const question = (state = [], { type, question }) => {
  const questionEquality = (a = {}, b = {}) => a.question_id === b.question_id;

  switch (type) {
    case 'FETCHED_QUESTION':
      return (state = unionWith([question], state, questionEquality));
    default:
      return state;
  }
};
