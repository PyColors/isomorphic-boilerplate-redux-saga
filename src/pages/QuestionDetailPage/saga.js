import { takeEvery, put } from 'redux-saga/effects';
import fetch from 'isomorphic-fetch';
import * as actions from './constants';

// Fetch question details from local proxy API
export function* handleFetchQuestion({ question_id }) {
  const raw = yield fetch(`/api/questions/${question_id}`);
  const json = yield raw.json();
  const question = json.items[0];

  // Question has been fetched
  yield put({
    type: actions.FETCHED_QUESTION,
    question
  });
}

export default function*() {
  yield takeEvery(actions.REQUEST_FETCH_QUESTION, handleFetchQuestion);
}
