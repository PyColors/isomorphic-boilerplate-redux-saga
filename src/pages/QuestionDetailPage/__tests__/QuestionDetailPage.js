import renderer from 'react-test-renderer';
import React from 'react';
import { mapStateToProps, QuestionDetailDisplay } from '../QuestionDetailPage';

describe('The Question Detail <Component', () => {
  describe('The Container Element', () => {
    /**
     * As Map State to Props is a pure function, it's lovely test it ðŸ˜Ž
     * without without adding libraries.
     */
    describe('mapStateToProps', () => {
      it('should map the state to props correctly', () => {
        const sampleQuestion = {
          question_id: 42,
          body: 'Lorem Ipsum'
        };
        const appState = {
          questions: [sampleQuestion]
        };
        const ownProps = {
          question_id: 42
        };
        const componentState = mapStateToProps(appState, ownProps);
        expect(componentState).toEqual(sampleQuestion);
      });
    });
  });

  describe('The display element', () => {
    it('Should not regress', () => {
      const tree = renderer.create(
        <QuestionDetailDisplay
          title="Lorem Ipsum"
          body="42"
          answerCount={0}
          tags={['react']}
        />
      );

      expect(tree.toJSON()).toMatchSnapshot();
    });
  });
});
