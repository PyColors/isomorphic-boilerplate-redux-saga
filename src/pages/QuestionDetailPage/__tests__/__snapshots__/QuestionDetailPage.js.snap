// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`The Question Detail <Component The display element Should not regress 1`] = `
<div>
  <h3
    className="mb-2"
  >
    Lorem Ipsum
  </h3>
  <div>
    <div
      className="mb-3"
    >
      <div>
        <code>
          react
        </code>
      </div>
    </div>
    <div>
      <p>
        42
      </p>
    </div>
    <div>
      0
       Answers
    </div>
  </div>
</div>
`;
